llvm-general/src/LLVM/General/Internal/Module.hs

```diff
@@ -217,17 +219,39 @@ getDataLayout m = do
   dlString <- decodeM =<< FFI.getDataLayout m
   either fail return . runExcept . parseDataLayout A.BigEndian $ dlString
 
--- | Build an LLVM.General.'Module' from a LLVM.General.AST.'LLVM.General.AST.Module' - i.e.
--- lower an AST from Haskell into C++ objects.
+
+genCodingInstance [t| Maybe A.G.UnnamedAddr |] ''FFI.UnnamedAddr [
+  (FFI.unnamedAddrNone, Nothing),
+  (FFI.unnamedAddrLocal, Just A.G.LocalAddr),
+  (FFI.unnamedAddrGlobal, Just A.G.GlobalAddr)
+ ]
+
+-- | This function will call disposeModule after the callback exits
 withModuleFromAST :: Context -> A.Module -> (Module -> IO a) -> ExceptT String IO a
-withModuleFromAST context@(Context c) (A.Module moduleId dataLayout triple definitions) f = runEncodeAST context $ do
-  moduleId <- encodeM moduleId
-  m <- anyContToM $ bracket (FFI.moduleCreateWithNameInContext moduleId c) FFI.disposeModule
-  maybe (return ()) (setDataLayout m) dataLayout
-  maybe (return ()) (setTargetTriple m) triple
+withModuleFromAST context@(Context c) mod f = runEncodeAST context $ do
+  moduleId <- encodeM (A.moduleName mod)
+  ffiMod <- anyContToM $ bracket (FFI.moduleCreateWithNameInContext moduleId c) FFI.disposeModule
+  astToFFIModule' mod ffiMod
+  liftIO $ f (Module ffiMod)
+
+astToFFIModule :: Context -> A.Module -> ExceptT String IO Module
```

Comment:
Yep, the linking API destroys modules so always calling disposemodule is no longer valid.